core/core.go

```diff
@@ -115,54 +140,110 @@ func (c *Core) InsertChain(blocks types.Blocks) (int, error) {
+	}
+	c.serviceBlocks(hashNumberList)
+	if len(hashNumberList) > 0 {
+		log.Info("Size of hashNumberList", "len", len(hashNumberList), "first entry", hashNumberList[0].Number, "last entry", hashNumberList[len(hashNumberList)-1].Number)
+	}
+}
+
+func (c *Core) serviceBlocks(hashNumberList []types.HashAndNumber) {
 	sort.Slice(hashNumberList, func(i, j int) bool {
 		return hashNumberList[i].Number < hashNumberList[j].Number
 	})
 
-	// Only take c_appendQueueThreshold latest blocks out of the database because we know that the
-	// append will be interrupted once we reach the dom block, so no need to get
-	// all the blocks in the appendQueue and load it to the RAM
-	var threshold int
-	if len(hashNumberList) > c_appendQueueThreshold {
-		threshold = c_appendQueueThreshold
-	} else {
-		threshold = len(hashNumberList)
+	var retryThreshold uint64
+	switch common.NodeLocation.Context() {
+	case common.PRIME_CTX:
+		retryThreshold = c_primeRetryThreshold
+	case common.REGION_CTX:
+		retryThreshold = c_regionRetryThreshold
+	case common.ZONE_CTX:
+		retryThreshold = c_zoneRetryThreshold
 	}
+
 	// Attempt to service the sorted list
-	for _, hashAndNumber := range hashNumberList[:threshold] {
-		block := c.GetBlockOrCandidateByHash(hashAndNumber.Hash)
-		if block != nil {
-			c.serviceFutureBlock(block)
+	for _, hashAndNumber := range hashNumberList {
+		header := c.GetHeaderOrCandidateByHash(hashAndNumber.Hash)
+		if header != nil {
+			var numberAndRetryCounter blockNumberAndRetryCounter
+			if value, exist := c.appendQueue.Peek(header.Hash()); exist {
+				numberAndRetryCounter = value.(blockNumberAndRetryCounter)
+				numberAndRetryCounter.retry += 1
+				if numberAndRetryCounter.retry > retryThreshold && numberAndRetryCounter.number+c_appendQueueRemoveThreshold < c.CurrentHeader().NumberU64() {
+					c.appendQueue.Remove(header.Hash())
+				} else {
+					c.appendQueue.Add(header.Hash(), numberAndRetryCounter)
+				}
+			}
+			parentHeader := c.GetHeader(header.ParentHash(), header.NumberU64()-1)
+			if parentHeader != nil {
+				// If parent header is dom, send a signal to dom to request for the block if it doesnt have it
+				_, parentHeaderOrder, err := c.sl.engine.CalcOrder(parentHeader)
+				if err != nil {
+					log.Warn("Error calculating the parent block order in serviceBlocks", "Hash", parentHeader.Hash(), "Number", parentHeader.NumberArray())
+				}
+				nodeCtx := common.NodeLocation.Context()
+				if parentHeaderOrder < nodeCtx && c.GetHeaderByHash(parentHeader.Hash()) == nil {
+					log.Info("Requesting the dom to get the block if it doesnt have and try to append", "Hash", parentHeader.Hash(), "Order", parentHeaderOrder)
+					// send a signal to the required dom to fetch the block if it doesnt have it, or its not in its appendqueue
+					go c.sl.domClient.RequestDomToAppendOrFetch(context.Background(), parentHeader.Hash(), parentHeaderOrder)
+				}
+				c.InsertChain([]*types.Block{types.NewBlockWithHeader(header)})
```

Comment:
We're inserting a block without transactions, should we verify that the state is not necessary?

Reply:
added a comment to explain, it is done for saving ram purpose, and using an empty block doesnt effect the append because we read the block form the database inside append